Firestore/core/src/firebase/firestore/util/async_queue.h

```diff
@@ -0,0 +1,161 @@
+// Operations may be scheduled to be executed as soon as possible or in the
+// future. Operations scheduled for the same time are FIFO-ordered.
+//
+// `AsyncQueue` wraps a platform-specific executor, adding checks that enforce
+// sequential ordering of operations: an enqueued operation, while being run,
+// normally cannot enqueue other operations for immediate execution, called
+// "nesting" (but see `EnqueueAllowingNesting`).
+//
+// `AsyncQueue` methods have particular expectations about whether they must be
+// invoked on the queue or not; check "preconditions" section in comments on
+// each method.
+//
+// A significant portion of `AsyncQueue` interface only exists for test purposes
+// and must *not* be used in regular code.
+class AsyncQueue {
+ public:
+  using Operation = internal::Executor::Operation;
+  // A more-or-less arbitrary unit of time for scheduling operations in the
+  // future.
+  using Milliseconds = internal::Executor::Milliseconds;
+
+  explicit AsyncQueue(std::unique_ptr<internal::Executor> executor);
+
+  // Asserts for the caller that it is being invoked asynchronously on the
+  // `AsyncQueue.`
+  void VerifyIsAsyncCall() const;
+  // Asserts for the caller that it is being invoked as part of an operation on
+  // the `AsyncQueue`.
+  void VerifyCalledFromOperation() const;
+
+  // Puts the `operation` on the queue to be executed as soon as possible, while
+  // maintaining FIFO order.
+  //
+  // Precondition: `Enqueue` calls cannot be nested; that is, `Enqueue` may not
+  // be called by a previously enqueued operation when it is run (as a special
+  // case, destructors invoked when an enqueued operation has run and is being
+  // destroyed may invoke `Enqueue`).
+  void Enqueue(const Operation& operation);
+
+  // Like `Enqueue`, but allowing nesting.
+  void EnqueueAllowingNesting(const Operation& operation);
+
+  // Puts the `operation` on the queue to be executed `delay` milliseconds from
+  // now, and returns a handle that allows to cancel the operation (provided it
+  // hasn't run already).
+  //
+  // `operation` is tagged by a `timer_id` which allows to identify the caller.
+  // Only one operation tagged with any given `timer_id` may be on the queue at
+  // any time; an attempt to put another such operation will result in an
+  // assertion failure. In tests, these tags also allow to check for presence of
+  // certain operations and to run certain operations in advance.
+  //
+  // Precondition: `EnqueueAfterDelay` is being invoked asynchronously on the
+  // queue.
+  DelayedOperation EnqueueAfterDelay(Milliseconds delay,
+                                     TimerId timer_id,
+                                     const Operation& operation);
+
+  // Immediately executes the `operation` on the queue.
+  //
+  // Precondition: the queue is idle at the moment of the call (no other
+  // operation is currently being executed).
```

Comment:
I was initially confused by this precondition until I read the implementation. Maybe add "i.e. no nesting" or similar to the comment?

(BTW, I also like the "nesting" term. It helps quite a bit.)

Reply:
Do you think the comment is confusing on its own terms, or is perhaps the issue that `StartExecution` is not clear enough? Perhaps rephrase to _"Precondition: no other operation is being executed on the queue at the moment of the call (i.e., calls to `StartExecution` cannot be nested)?"_ Otherwise, let's discuss in person tomorrow.

Reply:
Done.